src/mono/mono/eventpipe/ep-rt-mono.h

```diff
@@ -14,27 +18,32 @@
 #include <mono/utils/mono-time.h>
 #include <mono/utils/mono-rand.h>
 #include <mono/metadata/w32file.h>
-#include "ep.h"
+#include <mono/metadata/w32event.h>
 
 #undef EP_ARRAY_SIZE
 #define EP_ARRAY_SIZE(expr) G_N_ELEMENTS(expr)
 
 #undef EP_INFINITE_WAIT
 #define EP_INFINITE_WAIT MONO_INFINITE_WAIT
 
-//TODO: Should make sure block is executed in safe mode.
 #undef EP_GCX_PREEMP_ENTER
-#define EP_GCX_PREEMP_ENTER {
+#define EP_GCX_PREEMP_ENTER { MONO_REQ_GC_UNSAFE_MODE;
 
-//TODO: Should make sure block is returned back to previous mode.
 #undef EP_GCX_PREEMP_EXIT
-#define EP_GCX_PREEMP_EXIT }
+#define EP_GCX_PREEMP_EXIT MONO_REQ_GC_UNSAFE_MODE; }
```

Comment:
not sure why you need to check that we're in unsafe mode at exit.  is this so that you can be sure that there wasn't an unbalanced unsafe->safe transition?

Reply:
That was the intent, just making sure nothing incorrectly altered state in-between, but I guess we could turn it into a nop since that transition is currently just for CoreCLR since library running under Mono mainly operates in GC unsafe mode at the moment.